src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -527,29 +527,47 @@ bool IsDerivedType(TypeSymbol possibleDerived, TypeSymbol possibleBase)
-                    BoundPattern pattern = subPattern.Pattern;
-                    BoundDagEvaluation evaluation;
-                    switch (symbol)
+                    BoundPattern pattern = subpattern.Pattern;
+                    BoundDagTemp currentInput = input;
+                    if (subpattern.Symbols.IsEmpty)
                     {
-                        case PropertySymbol property:
-                            evaluation = new BoundDagPropertyEvaluation(pattern.Syntax, property, OriginalInput(input, property));
-                            break;
-                        case FieldSymbol field:
-                            evaluation = new BoundDagFieldEvaluation(pattern.Syntax, field, OriginalInput(input, field));
-                            break;
-                        default:
-                            RoslynDebug.Assert(recursive.HasAnyErrors);
-                            tests.Add(new Tests.One(new BoundDagTypeTest(recursive.Syntax, ErrorType(), input, hasErrors: true)));
-                            continue;
+                        RoslynDebug.Assert(recursive.HasAnyErrors);
+                        tests.Add(new Tests.One(new BoundDagTypeTest(recursive.Syntax, ErrorType(), input, hasErrors: true)));
+                        goto done;
                     }
 
-                    tests.Add(new Tests.One(evaluation));
-                    var element = new BoundDagTemp(pattern.Syntax, symbol.GetTypeOrReturnType().Type, evaluation);
-                    tests.Add(MakeTestsAndBindings(element, pattern, bindings));
+                    Symbol last = subpattern.Symbols.Last();
+                    foreach (Symbol symbol in subpattern.Symbols)
+                    {
+                        BoundDagEvaluation evaluation;
+                        switch (symbol)
+                        {
+                            case PropertySymbol property:
+                                evaluation = new BoundDagPropertyEvaluation(pattern.Syntax, property, OriginalInput(currentInput, property));
+                                break;
+                            case FieldSymbol field:
+                                evaluation = new BoundDagFieldEvaluation(pattern.Syntax, field, OriginalInput(currentInput, field));
+                                break;
+                            default:
+                                RoslynDebug.Assert(recursive.HasAnyErrors);
+                                tests.Add(new Tests.One(new BoundDagTypeTest(recursive.Syntax, ErrorType(), currentInput, hasErrors: true)));
+                                goto done;
+                        }
+
+                        tests.Add(new Tests.One(evaluation));
+                        TypeSymbol type = symbol.GetTypeOrReturnType().Type;
+                        currentInput = new BoundDagTemp(pattern.Syntax, type, evaluation);
+
+                        if (!ReferenceEquals(symbol, last))
```

Comment:
Is this way of checking whether we're reached the last symbol robust? You could have a pattern where the same symbol is involved more than once, no? Consider using a `for` loop instead.